src/runtime_src/hip/core/memory_pool.cpp

```diff
@@ -0,0 +1,407 @@
+      break;
+
+    case hipMemPoolAttrUsedMemCurrent:
+      m_used_mem_current = *reinterpret_cast<uint64_t*>(value);
+      break;
+
+    case hipMemPoolAttrUsedMemHigh:
+      m_used_mem_high = *reinterpret_cast<uint64_t*>(value);
+      break;
+    };
+  }
+
+  bool
+  memory_pool::extend_memory_list(size_t size)
+  {
+    m_list.insert(m_list.begin(), std::make_shared<memory_pool_node>(m_device, size, m_last_id++));
+    return true;
+  }
+
+  void memory_pool::malloc(void**ptr, size_t size)
+  {
+    *ptr = alloc(size);
+  }
+
+  void*
+  memory_pool::alloc(size_t size)
+  {
+    if (m_list.size() == 0) {
+      init();
+    }
+
+    size_t page_size = xrt_core::getpagesize();
+    size_t aligned_size = ((size + page_size) / page_size) * page_size;
+
+    std::lock_guard lock(m_mutex);
+
+    if (aligned_size > m_pool_size)
+    {
+      throw std::runtime_error("requested size is greater than memory pool block size.");
+    }
+
+    std::shared_ptr<memory_pool_node> mm;
+
+    uint32_t num_tries = 0;
+    // find first free slot that fits, if none is found, enlarge the pool and try one more time
+    while (num_tries <= 2) {
+      ++ num_tries;
+      auto mp_itr = m_list.begin();
+      while (mp_itr != m_list.end())
+      {
+        mm = *mp_itr;
+        if (m_pool_size - mm->m_used < aligned_size)
+        {
+          mp_itr++;
+          continue;
+        }
+
+        auto _free = mm->m_free_list;
+        auto _not_free = _free;
+
+        while (_free != nullptr)
+        {
+          if (_free->m_size >= aligned_size)
+          {
+            if (_free->m_size > aligned_size)
+            {
+              _not_free = _free;
+
+              _free = std::make_shared<memory_pool_slot>(_not_free->start() + aligned_size, _not_free->m_size - aligned_size);
+
+              // update free_list
+              if (_free->m_prev == nullptr) {
+                mm->m_free_list = _free;
+              }
+              else {
+                _free->m_prev->m_next = _free;
+              }
+              if (_free->m_next) {
+                _free->m_next->m_prev = _free;
+              }
+
+              _not_free->m_is_free = false;
+              _not_free->m_size = aligned_size;
+            }
+            else
+            {
+              _not_free = _free;
+              dlinkedlist_delete(mm->m_free_list, _not_free);
+              _not_free->m_is_free = false;
+            }
+            dlinkedlist_insert(mm->m_alloc_list, _not_free);
+
+            mm->m_used += _not_free->m_size;
+            m_used_mem_current += _not_free->m_size;
+            m_used_mem_high = m_used_mem_current;
+
+            size_t new_mem_start = reinterpret_cast<uint64_t>(mm->m_memory->get_address()) + _not_free->m_start;
+            return reinterpret_cast<void*>(new_mem_start);
+          }
+          _free = _free->m_next;
+        }
+        mp_itr++;
+        mm = *mp_itr;
+      }
+      
+      if (num_tries == 2) break;
```

Comment:
```suggestion
      if (num_tries == 2) 
        break;
```
Please fix all such occurences.

Reply:
Done.